hydra_base/lib/template/utils.py

```diff
@@ -0,0 +1,219 @@
+
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# (c) Copyright 2013 to 2017 University of Manchester
+#
+# HydraPlatform is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# HydraPlatform is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with HydraPlatform.  If not, see <http://www.gnu.org/licenses/>
+#
+
+import json
+import logging
+import re
+
+from hydra_base import db
+from hydra_base.db.model import Attr
+from hydra_base.lib.objects import JSONObject
+from hydra_base.exceptions import HydraError
+from hydra_base.util import dataset_util
+from hydra_base.lib import units
+
+log = logging.getLogger(__name__)
+
+def get_attr(attr_id):
+    attr = db.DBSession.query(Attr).filter(Attr.id == attr_id).one()
+    return JSONObject(attr)
+
+def check_dimension(typeattr, unit_id=None):
+    """
+        Check that the unit and dimension on a type attribute match.
+        Alternatively, pass in a unit manually to check against the dimension
+        of the type attribute
+    """
+    if unit_id is None:
+        unit_id = typeattr.unit_id
+
+    dimension_id = get_attr(typeattr.attr_id).dimension_id
+
+    if unit_id is not None and dimension_id is None:
+        # First error case
+        unit_dimension_id = units.get_dimension_by_unit_id(unit_id).id
+        unit = units.get_unit(unit_id)
+        dimension = units.get_dimension(unit_dimension_id, do_accept_dimension_id_none=True)
+        raise HydraError(f"Unit {unit_id} ({unit.abbreviation}) has dimension_id"+
+                         f" {dimension.id}(name=dimension.name),"+
+                         " but attribute has no dimension")
+    elif unit_id is not None and dimension_id is not None:
+        unit_dimension_id = units.get_dimension_by_unit_id(unit_id).id
+        unit = units.get_unit(unit_id)
+        dimension1 = units.get_dimension(unit_dimension_id, do_accept_dimension_id_none=True)
+        dimension2 = units.get_dimension(unit_dimension_id, do_accept_dimension_id_none=True)
+        if unit_dimension_id != dimension_id:
+            # Only error case
+            raise HydraError(f"Unit {unit_id} ({unit.abbreviation}) has dimension_id"+
+                             f" {dimension1.id}(name=dimension1.name),"+
+                             f" but attribute has id: {dimension2.id}({dimension2.name})")
```

Comment:
It should not be possible for an attribute (a type attribute i guess you mean?) to have a unit when its parent attribute has no dimension. A dataset / type attribute cannot have a unit unless the attribute has a dimension. Otherwise adding / updating these should fail.

Can you see somewhere where this is allowed?

Reply:
This is the case being tested for in 49 (then `raised` as an error) but the question is to how that case can arise - presumably this is possible by some route if it needs testing for here?  And if it's possible, but an error, can it prevented at creation rather than tested for later?

Reply:
That's a fair point, but surely it's better to check than to not check, even if checks are put in place at creation?

Reply:
on re-reading this code, this IS the check on creation. So if i set a unit ID on a typeattr, and try to add it, then this will catch that. 

This function is used by multiple different functions, both at creation and update.

Reply:
OK - all becomes clear...